### PR TITLE
Format and type specs

### PR DESCRIPTION
**Pull request by robinmonjo**
The goal of this PR is to use the Elixir 1.6 code formatter and write type specs on the entire codebase

So far I set up the code formatter. I had to:
- install Elixir 1.6 locally (obviously)
- install Elixir from source in Travis so I can run the formatter check
- get credo from the master branch to avoid this issue on Elixir 1.6: https://github.com/rrrene/credo/issues/463

TODOs:
- [x] ensure I follow Jose advices from https://github.com/elixir-lang/elixir/issues/6643 on how to handle formatting
- [x] write type specs !!
- [x] do this https://github.com/robinmonjo/coincoin/issues/18 type specs will be better
- [ ] <del>run dialyzer in the CI ? (this will make things slow or maybe Travis can cache PLT ?)</del> possible http://blog.danielberkompas.com/elixir/2015/04/03/run-dialyzer-on-elixir-on-travis.html but won't do it ...

**Once Elixir 1.6 is released I will have to "simplify" the `travis.yml`**

Fix: https://github.com/robinmonjo/coincoin/issues/14
Fix: https://github.com/robinmonjo/coincoin/issues/18

### CODE REVIEW

File: apps/blockchain/lib/blockchain/block.ex [hunk 1]
```diff
@@ -24,23 +35,24 @@ defmodule Blockchain.Block do
     }
   end
 
-  def generate_next_block(data) do
-    generate_next_block(data, Chain.latest_block)
-  end
+  @spec generate_next_block(any() | any(), t) :: t
```
Comment (robinmonjo) [inline apps/blockchain/lib/blockchain/block.ex]: idem `any()` here should refer to a type that implements `BlockData` protocol

**Comment by robinmonjo** [comment]
Ok I think I'm done with this PR. @yordis if you want to have a look, be my guest 😊

File: .formatter.exs [hunk 2]
```diff
@@ -0,0 +1,4 @@
+# Used by "mix format"
+[
+  inputs: ["mix.exs", "apps/*/mix.exs", "apps/*/{config,lib,test}/**/*.{ex,exs}"]
+]
```
Comment (yordis) [inline .formatter.exs]: Missing jump line.

Add editorconfig to the project

```
# .editorconfig
root = true

[*]
indent_style = space
indent_size = 2
end_of_line = lf
charset = utf-8
trim_trailing_whitespace = true
insert_final_newline = true

[*.md]
trim_trailing_whitespace = false
```

Most of the IDE and Text Editor support EditorConfig so you can avoid those details

**Comment by yordis** [comment]
Did you ran `elixir format`?

**Comment by robinmonjo** [comment]
Yes I did ran `mix format`, formatting is also checked by CI but I didn't include `.formatter.exs` in the list of file to format 😊. Fixed in last commit

**Comment by yordis** [comment]
Add `.editorconfig`, it will help with situations of trimming the spaces and adding the jump of line at the end of the file

```
# .editorconfig
root = true

[*]
indent_style = space
indent_size = 2
end_of_line = lf
charset = utf-8
trim_trailing_whitespace = true
insert_final_newline = true

[*.md]
trim_trailing_whitespace = false
```

**Comment by yordis** [comment]
Well you need to install the plugin for your editor for sure 😄

**Comment by robinmonjo** [comment]
Don't you think this would be redundant with `mix format` ? I think `mix format` already ensure everything you list in the config file

**Comment by yordis** [comment]
@robinmonjo does `mix format` add the jump at the end of the file or trim trailing whitespace? 

I think it doesn't because some of your files are missing the jump at the end of the line

Also is for all the files on the system

**Comment by robinmonjo** [comment]
Yes it does add the jumpline (only on files that matches the patterns defined in `.formatter.exs`. As for trailing whitespaces I'm not sure my editor does it for me. Will take a look at editor config (you are the first one who talk to me about this)

**Comment by yordis** [comment]
I lik editorconfig because you help the people that contributes to the project and you mitigate the issue with IDE/Editors configuration, at least they could install the plugin (which some of the editors and IDE already have it by default) and it will help us out.

File: .dockerignore [hunk 3]
```diff
@@ -1 +1,6 @@
-_build
\ No newline at end of file
+_build
+deps
+tmp
+.fetch
+*.ez
+erl_crash.dump
```
Comment (yordis) [inline .dockerignore]: @robinmonjo

File: Dockerfile [hunk 4]
```diff
@@ -1,11 +1,8 @@
...
+RUN apk update && apk add bash make git
 
 ADD . /app
 WORKDIR /app
-RUN mix local.hex --force && \
-    mix local.rebar --force && \
-    mix deps.get && \
-    MIX_ENV=prod mix release --env=prod
+RUN make release
 CMD _build/prod/rel/coincoin/bin/coincoin console
```
Comment (yordis) [inline Dockerfile]: @robinmonjo

File: .dialyzer_ignore [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+# https://elixirforum.com/t/elixir-1-6-dialyzer-genserver-issue/9982
+
+lib/agent.ex:162: The call 'Elixir.Supervisor':child_spec(_@2::#{'id':='Elixir.Blockchain.P2P.Peers', 'start':={'Elixir.Blockchain.P2P.Peers','start_link',[any(),...]}},[]) breaks the contract (child_spec() | {module(),arg::term()} | module(),elixir:keyword()) -> child_spec()
+lib/gen_server.ex:581: The call 'Elixir.Supervisor':child_spec(_@2::#{'id':='Elixir.Blockchain.Chain', 'start':={'Elixir.Blockchain.Chain','start_link',[any(),...]}},[]) breaks the contract (child_spec() | {module(),arg::term()} | module(),elixir:keyword()) -> child_spec()
+lib/gen_server.ex:581: The call 'Elixir.Supervisor':child_spec(_@2::#{'id':='Elixir.Blockchain.Mempool', 'start':={'Elixir.Blockchain.Mempool','start_link',[any(),...]}},[]) breaks the contract (child_spec() | {module(),arg::term()} | module(),elixir:keyword()) -> child_spec()
+lib/gen_server.ex:581: The call 'Elixir.Supervisor':child_spec(_@2::#{'id':='Elixir.Token.MyWallet', 'start':={'Elixir.Token.MyWallet','start_link',[any(),...]}},[]) breaks the contract (child_spec() | {module(),arg::term()} | module(),elixir:keyword()) -> child_spec()
```
Comment (yordis) [inline .dialyzer_ignore]: @robinmonjo

**Comment by robinmonjo** [comment]
Ok @yordis I added the `.editorconfig` file :+1:

**Comment by robinmonjo** [comment]
@yordis is it all good for you ? Should we merge this ?

**Comment by yordis** [comment]
@robinmonjo all good man! Amazing work 😎

**Comment by robinmonjo** [comment]
👍 thanks again for the review and advices ! I started to read the "Metaprogramming Elixir" book to see what can be done with the AST and how to use it to build a "blockchain language" for smart contracts. Lot's of work, big subject but might be interesting :)

**Pull request merged**